File: src/services/academic.ts
import { apiClient } from './api';
import { Course, Assignment, StudyGoal, AcademicData, SemesterStats } from '../features/academic/types';

export interface CreateCoursePayload {
  code: string;
  title: string;
  instructor: string;
  credits: number;
  color?: string;
}

export interface CreateAssignmentPayload {
  course_id: string;
  title: string;
  due_date: string;
  priority: 'high' | 'medium' | 'low';
  percent_of_grade: number;
}

export interface UpdateStudyGoalPayload {
  label: string;
  target_hours: number;
  completed_hours?: number;
}

export const academicService = {
  async getData(): Promise<AcademicData> {
    const { data } = await apiClient.get<AcademicData>('/academic/overview');
    return data;
  },

  async getStats(): Promise<SemesterStats> {
    const { data } = await apiClient.get<SemesterStats>('/academic/stats');
    return data;
  },

  async getCourses(): Promise<Course[]> {
    const { data } = await apiClient.get<Course[]>('/academic/courses');
    return data;
  },

  async createCourse(payload: CreateCoursePayload): Promise<Course> {
    const { data } = await apiClient.post<Course>('/academic/courses', payload);
    return data;
  },

  async updateCourse(id: string, payload: Partial<CreateCoursePayload & { current_grade: string; grade_point: number; progress: number }>): Promise<Course> {
    const { data } = await apiClient.patch<Course>(`/academic/courses/${id}`, payload);
    return data;
  },

  async deleteCourse(id: string): Promise<void> {
    await apiClient.delete(`/academic/courses/${id}`);
  },

  async getAssignments(filters?: { status?: string; course_id?: string }): Promise<Assignment[]> {
    const params = new URLSearchParams(filters as Record<string, string>);
    const { data } = await apiClient.get<Assignment[]>(`/academic/assignments?${params}`);
    return data;
  },

  async createAssignment(payload: CreateAssignmentPayload): Promise<Assignment> {
    const { data } = await apiClient.post<Assignment>('/academic/assignments', payload);
    return data;
  },

  async updateAssignment(id: string, payload: Partial<{ status: string; title: string }>): Promise<Assignment> {
    const { data } = await apiClient.patch<Assignment>(`/academic/assignments/${id}`, payload);
    return data;
  },

  async submitAssignment(id: string): Promise<Assignment> {
    const { data } = await apiClient.patch<Assignment>(`/academic/assignments/${id}/submit`);
    return data;
  },

  async getStudyGoals(): Promise<StudyGoal[]> {
    const { data } = await apiClient.get<StudyGoal[]>('/academic/study-goals');
    return data;
  },

  async upsertStudyGoal(payload: UpdateStudyGoalPayload): Promise<StudyGoal> {
    const { data } = await apiClient.post<StudyGoal>('/academic/study-goals', payload);
    return data;
  },

  async logStudyHours(goal_id: string, hours: number): Promise<StudyGoal> {
    const { data } = await apiClient.post<StudyGoal>(`/academic/study-goals/${goal_id}/log`, { hours });
    return data;
  },
};

File: backend/models/academic.py
"""Academic models: Course, Assignment, StudyGoal."""

import uuid
from datetime import datetime, date
from sqlalchemy import Column, String, Boolean, DateTime, Date, Text, Integer, Float, ForeignKey
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import relationship

from database import Base


class Course(Base):
    __tablename__ = "courses"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    code = Column(String(20), nullable=False)         # e.g. "CSC 301"
    title = Column(String(255), nullable=False)
    instructor = Column(String(255), nullable=True)
    credits = Column(Integer, default=3, nullable=False)

    # Grade tracking
    current_grade = Column(Float, nullable=True)      # percentage 0-100
    grade_point = Column(Float, nullable=True)        # 0.0 - 5.0 (or 4.0 depending on scale)

    # Progress percentage
    progress = Column(Integer, default=0, nullable=False)

    # UI theming
    color = Column(String(20), default="#4F46E5", nullable=False)

    # Status: active | completed | dropped
    status = Column(String(20), default="active", nullable=False)

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    user = relationship("User", back_populates="courses")
    assignments = relationship("Assignment", back_populates="course", cascade="all, delete-orphan")


class Assignment(Base):
    __tablename__ = "assignments"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    course_id = Column(UUID(as_uuid=True), ForeignKey("courses.id", ondelete="CASCADE"), nullable=False, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    title = Column(String(500), nullable=False)
    due_date = Column(DateTime, nullable=True)

    # Status: pending | submitted | graded | overdue
    status = Column(String(20), default="pending", nullable=False)

    # Priority: high | medium | low
    priority = Column(String(20), default="medium", nullable=False)

    # Weight in course grade (0-100)
    percent_of_grade = Column(Float, nullable=True)

    score = Column(Float, nullable=True)              # actual score if graded

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    course = relationship("Course", back_populates="assignments")
    user = relationship("User", back_populates="assignments")


class StudyGoal(Base):
    __tablename__ = "study_goals"

    id = Column(UUID(as_uuid=True), primary_key=True, default=uuid.uuid4, index=True)
    user_id = Column(UUID(as_uuid=True), ForeignKey("users.id", ondelete="CASCADE"), nullable=False, index=True)

    label = Column(String(255), nullable=False)          # e.g. "Study 10 hrs/week"
    target_hours = Column(Float, default=10.0, nullable=False)
    completed_hours = Column(Float, default=0.0, nullable=False)
    week_start = Column(Date, default=date.today, nullable=False)  # ISO Monday of the target week

    created_at = Column(DateTime, default=datetime.utcnow, nullable=False)
    updated_at = Column(DateTime, default=datetime.utcnow, onupdate=datetime.utcnow, nullable=False)

    # Relationships
    user = relationship("User", back_populates="study_goals")
